File: robotBoundedInCircle/python/robot_bounded_in_circle.py
class Solution:
    """
    Complexity
    ----------
    - TC: O(n)
    - SC: 0(1)
    """
    def isRobotBounded(self, instructions: str) -> bool:
        # north -> east -> south -> west
        directions = ((0, 1), (1, 0), (0, -1), (-1, 0))

        x = y = 0
        index = 0

        for instruction in instructions:
            if instruction == "R":
                index = (index + 1) % 4
            elif instruction == "L":
                index = (index + 3) % 4
            else:
                x += directions[index][0]
                y += directions[index][1]

        # back to the original point or not face north
        return (x == 0 and y == 0) or index != 0


if __name__ == "__main__":
    instructions = "GLRLLGLL"
    print(f"The solution to {instructions} is {Solution().isRobotBounded(instructions)}")

File: problems/robotBoundedInCircle/c++/robot_bounded_in_circle.cpp
#include<iostream>

using namespace std;

class Solution {
/*
Complexity
----------
- TC: O(n)
- SC: O(1)
*/
public:
    bool isRobotBoundedInCircle(string instructions) {
        int x = 0, y = 0, direction = 0;

        for (char c : instructions) {
            if (c == 'G') {
                go(direction, x, y);
            }
            else if (c == 'R') {
                direction = (direction + 90) % 360;
            }
            else if (c == 'L') {
                direction = ((direction - 90) + 360) % 360;
            }
        }
        return (x == 0 and y == 0) or (direction != 0);
    }

    void go(int direction, int& x, int& y) {
        if (direction == 0) {
            y ++;
        }
        else if (direction == 90) {
            x ++;
        }
        else if (direction == 180) {
            y --;
        }
        else if (direction == 270) {
            x --;
        }
    }
};

int main() {
    string instructions = "GLRLLGLL";
    Solution obj;
    cout << obj.isRobotBoundedInCircle(instructions) << endl;
}
